### ISSUE TITLE
issues in 2hop/3hop QA

### ISSUE THREAD

**Issue by Wangyinquan** [malllabiisc/EmbedKGQA#8]
I've trained a 200-dim ComplEx embedding for the knowledge graph of MetaQA.
As the best hyperparameters are not given, I directly use the default parameters. 
The EmbedKGQA model works well in 1hop QA and gets 0.957 accuracy which is not too far from the value 0.975 given in the paper.
However when I tried 2hop and 3hop QA, I got 0.09 and 0.326 respectively. It seems to be more than just a issue of hyperparameters.

I find that in the test set of 2hop and 3hop MetaQA, most triples (topic entity, some relation, answer entity) don't appear in the knowledge graph kb.txt ( 4.97MB ), especially 2hop dataset ( 14872 QAs' in test set ). In 2hop set there are only 8 questions which have link between topic entity and answer entity in kb.txt. Also, the overlap (topic entity, some question, answer entity) between test and train set is only a small fraction of test set (less than a half).
Am I using the wrong dataset? Did I miss some important details?

This situation is similar with that described in Section 5.4: there is no link between head entity and answer entity. I dont understand how it could be possible that got 0.299 accuracy in 1hop( Table 4 in your paper) but 0.988 in 2hop( Table 2 in your paper).

**Comment by apoorvumang** [comment]
Hi
As you have pointed out, in 2-hop test set there are only 8 questions which have link between topic entity and answer entity in kb.txt. However all have a 2-length path - this means that (topic entity, some relation, x) and (x, some other relation, answer entity) triples are present in kb. Direct link need not be present to answer question.
I am not sure about why you are getting bad results for 2 and 3 hop. Perhaps you could share with me the exact code you are running, and I can debug? 
Also in section 5.4 we are doing an ablation study by knowingly dropping certain links, and the results can't be compared with full kb 2-hop.

**Comment by apoorvumang** [comment]
Also if you are looking for cleaned up code (rather than the outline we have right now), I would request you to wait till Aug 5 since I have some commitments till then. I will be fully available after that.

**Comment by Wangyinquan** [comment]
Hi Apoorv, 
I'm glad to know that you'll upload cleaned up code after Aug 5, hope to see that soon.
I am confused that after dropping links in section 5.4, 1-hop QA becomes multi-hop QA, the model get 0.299 accuracy while 2-hop and 3-hop QA are originally multi-hop QA but get a high accuracy.
For 2-hop QA, more than a half of questions in test set have following properties:
    1)multi-hop
    2)have no direct link in KB
    3)have no paraphrases of the same question in train set.
I think they are the same settings as the ablation study in section 5.4, the model should not be able to solve it properly, But 0.988 accuracy induces that the model can answer almost all those questions. I wonder what's wrong with my view.
Also I'd like to make sure if you use RelScore or not in MetaQA. And did you use inverse relations in training KG embeddings?
Thanks a lot

**Comment by Hy-Choi** [comment]
Hi. @Wangyinquan 

I am training meta qa 1-hop data. But, Accuracy is very low. 
Can you sharing 1-hop model or training parameter?

**Comment by Wangyinquan** [comment]
Hi, @Hy-Choi 
To train the QA model, you should get a KG embedding for kb.txt of MetaQA before that and use the same model (ComplEx) as in EmbedKGQA. In my experience, HIT@1 gets almost 1 since KGE overfits on kb.txt. Then I train EmbedKGEQ with default parameters.

**Comment by zmLeung** [comment]
It seems that you've tried many different combinations of modules during the experiment. Can you make public the results of different combinations?

**Comment by Wangyinquan** [comment]
Hi, @zmLeung 
I just tried with default hyperparameters and found this issue. Maybe I did something wrong. Hope @apoorvumang will upload cleaned up code tomorrow.

**Comment by apoorvumang** [comment]
Please check out the updated code @Wangyinquan @zmLeung @Hy-Choi

**Comment by apoorvumang** [comment]
Closing this for now, I'm hoping the updated code will fix the problem. If not, please reopen this issue.
